servicetalk-client-api/src/main/java/io/servicetalk/client/api/DefaultAutoRetryStrategyProvider.java

```diff
@@ -0,0 +1,159 @@
+         * {@link RetryableException}. By default, automatic retries always retries these {@link RetryableException}s.
+         * This method disables the default behavior.
+         *
+         * @return {@code this}.
+         */
+        public Builder disableRetryAllRetryableExceptions() {
+            retryAllRetryableExceptions = false;
+            return this;
+        }
+
+        /**
+         * Updates maximum number of automatic retries done for any request.
+         *
+         * @param maxRetries Maximum number of automatic retries done for any request.
+         * @return {@code this}.
+         */
+        public Builder maxRetries(int maxRetries) {
+            if (maxRetries <= 0) {
+                throw new IllegalArgumentException("maxRetries " + maxRetries + " (expected >0)");
+            }
+            this.maxRetries = maxRetries;
+            return this;
+        }
+
+        /**
+         * Builds a new {@link AutoRetryStrategyProvider}.
+         *
+         * @return A new {@link AutoRetryStrategyProvider}.
+         */
+        public AutoRetryStrategyProvider build() {
+            return new DefaultAutoRetryStrategyProvider(maxRetries, waitForLb, retryAllRetryableExceptions);
+        }
+    }
+
+    private static final class DefaultAutoRetryStrategy implements AutoRetryStrategy {
+        @Nullable
+        private final LoadBalancerReadySubscriber loadBalancerReadySubscriber;
+        private final AsyncCloseable closeAsync;
+        private final int maxRetryCount;
+        private final boolean retryAllRetryableExceptions;
+
+        DefaultAutoRetryStrategy(final int maxRetryCount, final boolean waitForLb,
+                                 final boolean retryAllRetryableExceptions, final LoadBalancer<?> loadBalancer) {
+            this.maxRetryCount = maxRetryCount;
+            this.retryAllRetryableExceptions = retryAllRetryableExceptions;
+            if (waitForLb) {
+                loadBalancerReadySubscriber = new LoadBalancerReadySubscriber();
+                closeAsync = toAsyncCloseable(__ -> {
+                    loadBalancerReadySubscriber.cancel();
+                    return completed();
+                });
+                toSource(loadBalancer.eventStream()).subscribe(loadBalancerReadySubscriber);
+            } else {
+                loadBalancerReadySubscriber = null;
+                closeAsync = emptyAsyncCloseable();
+            }
+        }
+
+        @Override
+        public Completable apply(final int count, final Throwable cause) {
+            if (count > maxRetryCount) {
+                return failed(cause);
+            }
+            if (loadBalancerReadySubscriber != null && cause instanceof NoAvailableHostException) {
+                return loadBalancerReadySubscriber.onHostsAvailable();
+            }
+            if (retryAllRetryableExceptions && cause instanceof RetryableException) {
```

Comment:
Should we retry all `RetryableException`s by default? It will also retry `io.servicetalk.client.api.ConnectTimeoutException` that may cause very high latency.

Reply:
Are you assuming that all subsequent connects also will timeout?

For a degenerate case, when all servers fail connect then yes it would cause additional latency but otherwise client-side load balancing is expected to move off such "faulty hosts"